worker-macros/src/durable_object.rs

```diff
@@ -109,24 +124,123 @@ pub fn expand_macro(tokens: TokenStream) -> syn::Result<TokenStream> {
+                                // SAFETY:
+                                // On the surface, this is unsound because the Durable Object could be dropped
+                                // while JavaScript still has possession of the future. However,
+                                // we know something that Rust doesn't: that the Durable Object will never be destroyed
+                                // while there is still a running promise inside of it, therefore we can let a reference
+                                // to the durable object escape into a static-lifetime future.
+                                let static_self: &'static mut Self = unsafe {&mut *(self as *mut _)};
+
+                                wasm_bindgen_futures::future_to_promise(async move {
+                                    static_self._on_message_raw(ws.into(), message).await.map(|_| wasm_bindgen::JsValue::NULL)
+                                        .map_err(wasm_bindgen::JsValue::from)
+                                })
+                            }
+
+                            #method
+                        }
+                    },
+                    "on_close" => {
+                        optional_methods.has_on_close = true;
+
+                        let mut method = impl_method.clone();
+                        method.sig.ident = Ident::new("_on_close_raw", method.sig.ident.span());
+                        method.vis = Visibility::Inherited;
+
+                        quote! {
+                            #pound[wasm_bindgen::prelude::wasm_bindgen(js_name = webSocketClose)]
+                            pub fn _on_close(&mut self, ws: worker_sys::web_sys::WebSocket, code: usize, reason: String, was_clean: bool) -> js_sys::Promise {
+                                // SAFETY:
+                                // On the surface, this is unsound because the Durable Object could be dropped
+                                // while JavaScript still has possession of the future. However,
+                                // we know something that Rust doesn't: that the Durable Object will never be destroyed
+                                // while there is still a running promise inside of it, therefore we can let a reference
+                                // to the durable object escape into a static-lifetime future.
+                                let static_self: &'static mut Self = unsafe {&mut *(self as *mut _)};
+
+                                wasm_bindgen_futures::future_to_promise(async move {
+                                    static_self._on_close_raw(ws.into(), code, reason, was_clean).await.map(|_| wasm_bindgen::JsValue::NULL)
+                                        .map_err(wasm_bindgen::JsValue::from)
+                                })
+                            }
+
+                            #method
+                        }
```

Comment:
```suggestion
quote! {
                            #pound[wasm_bindgen::prelude::wasm_bindgen(js_name = webSocketMessage)]
                            pub async fn _on_message(&mut self, ws: worker_sys::web_sys::WebSocket, message: String) -> Result<wasm_bindgen::JsValue, worker::Error> {
                                    self._on_message_raw(ws.into(), message).await.map(|_| wasm_bindgen::JsValue::NULL)
                            }

                            #method
                        }
```

According to the `wasm_bindgen` docs it can automatically convert async function to `js_sys::Promise` and can also automatically handle returns types of `Result<T: Into<JsValue>, T: Into<JsValue>>` so we can remove 90% of this boilerplate and the `static_self` "unsafe" rebind.

Reply:
Further to this, you could implement `describe::WasmDescribe`, `convert::IntoWasmAbi`, and `convert::FromWasmAbi` by delegating to the inner implementations of the inner `web_sys::WebSocket` type like I did [here](https://github.com/rustwasm/wasm-bindgen/issues/2407) for Enums which would then basically remove the need for this macro all together I think as you wouldn't need the boilerplate of converting back and forth from the raw `web_sys` types and the wrapped `worker` types.